src/examples/mnist/MNIST_SP.cpp

```diff
@@ -24,210 +24,173 @@
       /* proximalSegmentThreshold */      3, // 7, // 14,
       /* proximalIncrement */             .032,
       /* proximalDecrement */             .00928,
       /* proximalSynapseThreshold */      .422,
-      /* distalMaxSegments */             0,
-      /* distalMaxSynapsesPerSegment */   0,
-      /* distalSegmentThreshold */        0,
-      /* distalSegmentMatch */            0,
-      /* distalAddSynapses */             0,
-      /* distalIncrement */               0,
-      /* distalDecrement */               0,
-      /* distalMispredictDecrement */     0,
-      /* distalSynapseThreshold */        0,
-      /* stability_rate */                0,
-      /* fatigue_rate */                  0,
+
+      /* distalMaxSegments */             0u,
+      /* distalMaxSynapsesPerSegment */   0u,
+      /* distalSegmentThreshold */        0u,
+      /* distalSegmentMatch */            0u,
+      /* distalAddSynapses */             0u,
+      /* distalIncrement */               (Permanence)0.0,
+      /* distalDecrement */               (Permanence)0.0,
+      /* distalMispredictDecrement */     (Permanence)0.0,
+      /* distalSynapseThreshold */        (Permanence)0.0,
+
+      /* stability_rate */                0.0,
+      /* fatigue_rate */                  0.0,
       /* period */                        1402,
       /* seed */                          0,
       /* verbose */                       verbosity);
 
-  SDR cells( htm.cellDimensions );
-  SDR_Metrics columnStats(cells, 1402);
+//  columns.initialize({sp.getNumColumns()});
+  columns.initialize(cp.cellDimensions);
 
-  SDRClassifier clsr(
+  clsr.initialize(
     /* steps */         {0},
     /* alpha */         .001,
     /* actValueAlpha */ .3,
                         verbosity);
 
+  dataset = mnist::read_dataset<std::vector, std::vector, uint8_t, uint8_t>(string("../ThirdParty/mnist_data/mnist-src/")); //from CMake
+}
+
+void train() {
   // Train
-  auto train_images = read_mnist_images("./mnist_data/train-images-idx3-ubyte");
-  auto train_labels = read_mnist_labels("./mnist_data/train-labels-idx1-ubyte");
+
   if(verbosity)
-    cout << "Training for " << (train_dataset_iterations * train_labels.size())
+    cout << "Training for " << (train_dataset_iterations * dataset.training_labels.size())
          << " cycles ..." << endl;
-  for(auto i = 0u; i < train_dataset_iterations; i++) {
+  size_t i = 0;
+
+  SDR_Metrics inputStats(input,    1402);
+  SDR_Metrics columnStats(columns, 1402);
+
+  for(auto epoch = 0u; epoch < train_dataset_iterations; epoch++) {
+    NTA_INFO << "epoch " << epoch;
     // Shuffle the training data.
-    vector<UInt> index( train_labels.size() );
-    for(auto s = 0u; s < train_labels.size(); s++)
-        index[s] = s;
-    Random(3).shuffle( index.begin(), index.end() );
+    vector<UInt> index( dataset.training_labels.size() );
+    index.assign(dataset.training_labels.cbegin(), dataset.training_labels.cend());
```

Comment:
The problem is here.  The list `index` contains an index into the training dataset.  It is shuffled before each dataset iteration.

What was:
```C++
// Create a list of the numbers [0 - trainingData.size()]
for(auto s = 0u; s < train_labels.size(); s++)
     index[s] = s;
```
Became some thing different:
```C++
// Creates a list of the data-labels for each datum in the training dataset.
index.assign(dataset.training_labels.cbegin(), dataset.training_labels.cend());
```
And it should probably changed to:
`std::iota`

Reply:
so it trained on mismatched data/labels.

Reply:
`index` should contain every index of every datum in the training dataset.

Instead it contained the numbers [0-9], which are the labels of the data, not the index of the data.